sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc

```diff
@@ -0,0 +1,795 @@
+trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc.
+used by permission by Khronos.
+
+== Status
+
+Working Draft
+
+This is a preview extension specification, intended to provide early access to
+a feature for review and community feedback. When the feature matures, this
+specification may be released as a formal extension.
+
+Because the interfaces defined by this specification are not final and are
+subject to change they are not intended to be used by shipping software
+products.
+
+== Version
+
+Built On: 2021-08-01 +
+Revision: 2
+
+== Contact
+
+Artem Radzikhovskyy, Intel (artem 'dot' radzikhovskyy 'at' intel 'dot' com)
+
+== Contributors
+
+Artem Radzikhovskyy, Intel +
+Michael Kinsner, Intel +
+Jessica Davies, Intel +
+Joe Garvey, Intel +
+Mohammad Fawaz, Intel +
+Tommy Hoffner, Intel +
+John Pennycook, Intel +
+Greg Lueck, Intel +
+Roland Schulz, Intel
+
+== Dependencies
+
+This extension is written against the SYCL 2020 specification, revision 3.
+
+It also depends on the `SYCL_EXT_ONEAPI_PROPERTY_LIST` extension.
+
+== Overview
+
+[NOTE]
+====
+In this document, we use `device_global` to indicate the proposed `sycl::ext::oneapi::device_global`.
+====
+
+The purpose of this document is to clearly describe and specify `device_global` and related
+concepts, types, and mechanisms, and to give examples and context for their usage.
+
+=== Motivation
+
+Device scope memory allocations can provide an efficient mechanism for communication
+between multiple invocations of a kernel, or between kernels executing on a device.
+There are additional benefits and optimization opportunities when a device compiler
+has visibility into the allocation size (static sizing) and uses of the allocation.
+
+Syntax allowing direct use of an allocation (without passing pointers or parameters
+through function call boundaries) can also lead to syntax simplification in some
+important use cases.
+
+=== Examples
+
+Two example `device_global` can be declared at namespace scope, as follows:
+
+[source,c++]
+----
+struct MyClass {
+  bool flag;
+};
+
+using namespace sycl::ext::oneapi;
+
+device_global<MyClass> dm1;
+static device_global<int[4]> dm2;
+----
+
+`dm1` creates an allocation on each `sycl::device` that contains an object of type `MyClass`.
+`dm2` has internal linkage (due to `static`), and creates allocations containing an array
+of four `int` per device.
+
+Uses of `dm1` and `dm2` in device functions are syntactically similar to uses of global variables
+in {cpp} (access directly through the namespace scope identifier), and `device_global` has
+reference wrapper-like semantics on a device.  Of note, because {cpp} doesn't allow for
+overloading of the "dot operator", a `get()` member of `device_global` allows a reference
+to be extracted, to which the usual dot operator may be applied as in:
+
+[source,c++]
+----
+sycl::queue Q;
+Q.submit([&](sycl::handler& h) {
+  h.single_task([=]() {
+    int x = 5;
+    if (dm1.get().flag)
```

Comment:
`operator->` is overloaded such that if the `device_global` contains a pointer to a type where `operator->` is valid, then `->` will work as expected.  `device_global` has similar semantics to a reference wrapper, and since the dot operator can't be overloaded, we end up needing `.get()` if the `device_global` contains a struct directly (not pointer to it).

Reply:
Great! That makes sense.
@erichkeane @AaronBallman do you remember where the dot operator got distracted in ISO C++? Just the overflow of more urgent proposals?

Reply:
@keryell -- It didn't have consensus to proceed under the form proposed. I wouldn't be surprised if there were additional proposals in the future, but AFAIK, nothing is planned.

Reply:
@AaronBallman thanks for the feedback. Too bad it got stuck.